atoMEC/postprocess/localization.py

```diff
@@ -0,0 +1,354 @@
+        elif spindims == 2:
+            xargs_min[0].append(-1)
+            xargs_min[1].append(-1)
+
+        return xargs_min
+
+    @staticmethod
+    def calc_N_shell(xargs_min, xgrid, density, ELF):
+        """
+        Compute the number of electrons in each shell from the ELF.
+
+        The number of electrons per shell is determined by integrating
+        the electron density between successive minima of the ELF.
+
+        Parameters
+        ----------
+        xargs_min : list
+            list of minima, as well as the first and last points
+        xgrid : ndarray
+            the logarithmic grid
+        density : ndarray
+            the electron density
+        ELF : ndarray
+            the electron localization function
+
+        Returns
+        -------
+        N_shell : list of floats
+            list of the number of electrons per shell
+        """
+        spindims = len(xargs_min)
+        N_shell = xargs_min  # initialize N_shell with the right shape
+
+        for i in range(spindims):
+            for j in range(len(xargs_min[i]) - 1):
+
+                # determine the part of the xgrid lying between two minima
+                xgrid_partition = xgrid[xargs_min[i][j] : xargs_min[i][j + 1]]
+
+                # determine the part of the density lying between two minima
+                density_partition = density[i][xargs_min[i][j] : xargs_min[i][j + 1]]
+
+                # this will be going...
+                # ELF_partition = ELF[i][xargs_min[i][j] : xargs_min[i][j + 1]]
+
+                # integrate over the density between two minima
+                N_shell[i][j] = mathtools.int_sphere(density_partition, xgrid_partition)
+
+        # delete the last element in N_shell (superfluous)
+        for i in range(spindims):
+            N_shell[i].pop()
+
+        return N_shell
+
+
+def calc_IPR_mat(eigfuncs, xgrid):
+    r"""
+    Calculate the inverse participation ratio for all eigenfunctions (see notes).
+
+    Parameters
+    ----------
+    eigfuncs : ndarray
+        transformed radial KS orbitals :math:`P_{nl}(x)=\exp(x/2)X_{nl}(x)`
+    xgrid : ndarray
+        the logarithmic grid
+
+    Returns
+    -------
+    IPR_mat : ndarray
+        the matrix of all IPR values
+
+    Notes
+    -----
+    The inverse participation ratio for a state `i` is usually defined as
+
+    .. math:: \mathrm{IPR}_i = \int \mathrm{d}{\mathbf{r}} |\Psi_i(\mathbf{r})|^4
+
+    It is typically used as a localization measure.
+
+    WARNING: the current definition in this version is not mathematically correct.
+    It does not include the proper contribution from the spherical harmonics
+    :math:`|Y_l^m(\theta,\phi)|^4`. This is omitted because it makes little difference
+    to the flavour of the results but complicates things. Currently, this function
+    does not correctly produce the expected limits (even if the spherical harmonic
+    contribution is correctly accounted for). Use at your own peril...
```

Comment:
![grafik](https://user-images.githubusercontent.com/28047702/157018752-41942443-e222-4fb4-b7b0-5cde2e584189.png)

```suggestion
    .. warning::
       The current definition in this version is not mathematically correct.
       It does not include the proper contribution from the spherical harmonics
       :math:`|Y_l^m(\theta,\phi)|^4`. This is omitted because it makes little difference
       to the flavour of the results but complicates things. Currently, this function
       does not correctly produce the expected limits (even if the spherical harmonic
       contribution is correctly accounted for). Use at your own peril...

```
![grafik](https://user-images.githubusercontent.com/28047702/157018818-c067691f-a574-4795-a778-3f6c09e322f4.png)